### DISCUSSION TITLE
Docker Deployment, how do the img css and js static files get deployed?

### DISCUSSION THREAD

**Discussion by kladderadeng** [mattieserver/netbox-topology-views#527] [Q&A] [answered]
I am running Netbox 4.0.6 in Docker and today I tried to get netbox-topology-views (4.0.0) running. I followed the mixed instructions here and on the netbox site to implement plugins in the container. So in the end I built my customized image that "seems" to run fine. On the first startup I could see, that the additional tables got created and I see the Topology Plugin in Netbox.
However, no Topologies are shown. So I looked into the logs and I can see 404s accessing img / css static folders:

```
netbox               | 172.20.0.10 - - [14/Jul/2024:07:04:31 +0000] "GET /plugins/netbox_topology_views/topology/?show_cables=True&show_power=True&group_sites=True&group_locations=True&region_id=1&site_id=1&role_id=2 HTTP/1.1" 200 142467 "https://netbox.example.com/plugins/netbox_topology_views/topology/?role_id=2&show_cables=True&draw_init=False" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0"
netbox               | 172.20.0.10 - - [14/Jul/2024:07:04:31 +0000] "GET /static/netbox_topology_views/css/app.css HTTP/1.1" 404 54 "https://netbox.example.com/plugins/netbox_topology_views/topology/?show_cables=True&show_power=True&group_sites=True&group_locations=True&region_id=1&site_id=1&role_id=2" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0"
netbox               | 172.20.0.10 - - [14/Jul/2024:07:04:31 +0000] "GET /static/netbox_topology_views/css/vendor.css HTTP/1.1" 404 54 "https://netbox.example.com/plugins/netbox_topology_views/topology/?show_cables=True&show_power=True&group_sites=True&group_locations=True&region_id=1&site_id=1&role_id=2" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0"
netbox               | 172.20.0.10 - - [14/Jul/2024:07:04:31 +0000] "GET /static/netbox_topology_views/js/app.js HTTP/1.1" 404 54 "https://netbox.example.com/plugins/netbox_topology_views/topology/?show_cables=True&show_power=True&group_sites=True&group_locations=True&region_id=1&site_id=1&role_id=2" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0"
```

I docker exec into the netbox container and had a look at the /opt/netbox/netbox/static/netbox_topology_views folder and the only folder that is there is the img folder (it was part of the Dockerfile-Plugins build process, but it is empty. Also the css and js folders are missing. I could now bind mount those folders and populate them by downloading the content from GitHub manually, but I think that will cause issues in the future when the plugin changes. So I wonder where in the process of activating the plugin, those folders should be populated and by whom?

**Comment by mattieserver** [comment, marked as answer]
The only steps you need to follow are at https://github.com/netbox-community/netbox-docker/wiki/Using-Netbox-Plugins.
When building the container the files are copied to the correct location.

You can try to fix it by entering the container and running:
`python3 manage.py collectstatic --no-input`

But this will only work until you re-create the container.

**Reply by kladderadeng** [reply]
Thank you very much for your response. I tried to do the python3 manage.py collectstatic --no-input before, but ran into permission issues.
However, now that you mention the Link regarding the "Using Netbox Plugins" instructions again, I modified that one. I thought that some parts were meant for the example used in the documentation which I thought refers if you want to install the secrets plugin. So my Dockerfile-Plugins looks different at the moment:

```
FROM netboxcommunity/netbox:latest

COPY ./plugin_requirements.txt /opt/netbox/
RUN /opt/netbox/venv/bin/pip install  --no-warn-script-location -r /opt/netbox/plugin_requirements.txt

# For the Netbox Topology Views Plugin
RUN mkdir -p /opt/netbox/netbox/static/netbox_topology_views/img
```

I will adjust it according to the example and include the secrets plugins. It did not cross my mind that this might be the base for other plugins, so I will start over and create a new image, Database etc. next.

**Reply by kladderadeng** [reply]
I think I got it now. After following the [example](https://github.com/netbox-community/netbox-docker/wiki/Using-Netbox-Plugins) and including the secrets Plugin for now, while creating the image I got a Python error regarding a missing module. Quick research revealed, that the secrets Plugin is not available for Netbox 4.x yet. So I ended up with the following changes to get the Docker Image created and having a working Topology plugin:

File: Dockerfile-Plugins
```
FROM netboxcommunity/netbox:latest

COPY ./plugin_requirements.txt /opt/netbox/
RUN /opt/netbox/venv/bin/pip install  --no-warn-script-location -r /opt/netbox/plugin_requirements.txt

# For the Netbox Topology Views Plugin
RUN mkdir -p /opt/netbox/netbox/static/netbox_topology_views/img

# These lines are only required if your plugin has its own static files.
COPY configuration/configuration.py /etc/netbox/config/configuration.py
COPY configuration/plugins.py /etc/netbox/config/plugins.py
RUN SECRET_KEY="dummydummydummydummydummydummydummydummydummydummy" /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py collectstatic --no-input
```

File: plugin_requirements.txt
```
netbox-topology-views
```

File: ./configuration/plugins.py
```
PLUGINS = ["netbox_topology_views"]
```

What got me in the beginning was the comment "# These lines are only required if your plugin has its own static files." in the instructions. Because there was a dummy SECRET_KEY and the example is an example of installing the secrets Plugin made me connect those two and decide. I do not want the secrets plugin, so the line with the SECRET_KEY can be left out.

This was all a beginners mistake and now I understand how the process works. Thank you very much for your patience. It is all working now :-)
